test/evp_kdf_test.c

```diff
@@ -160,6 +162,180 @@ static int test_kdf_pbkdf2(void)
+        && TEST_mem_eq(out2, sizeof(out2), expected_res2, sizeof(expected_res2));
+
+    if (ret == 0)
+        return 0;
+
+    EVP_KDF_free_ctx(kctx);
+
+# if defined(OPENSSL_THREADS)
+    /*
+     * Test threaded Argon2 variant. Note that this requires that threading
+     * was enabled during build.
+     */
+
+    CRYPTO_THREAD_enable(NULL, -1);
+    if (CRYPTO_THREAD_enabled(NULL) == 0) {
+        return ret;
+    }
+
+    static const unsigned char expected_res3[sizeof(out3)] = {
+        0xa8, 0x6c, 0x83, 0xa1, 0x9f, 0x0b, 0x23, 0x4e,
+        0xcb, 0xa8, 0xc2, 0x75, 0xd1, 0x6d, 0x05, 0x91,
+        0x53, 0xf9, 0x61, 0xe4, 0xc3, 0x9e, 0xc9, 0xb1,
+        0xbe, 0x98, 0xb3, 0xe7, 0x3d, 0x79, 0x17, 0x89,
+        0x36, 0x36, 0x82, 0x44, 0x3a, 0xd5, 0x94, 0x33,
+        0x40, 0x48, 0x63, 0x4e, 0x91, 0xc4, 0x93, 0xaf,
+        0xfe, 0xd0, 0xbc, 0x29, 0xfd, 0x32, 0x9a, 0x0e,
+        0x55, 0x3c, 0x00, 0x14, 0x9d, 0x6d, 0xb1, 0x9a,
+        0xf4, 0xe4, 0xa3, 0x54, 0xae, 0xc1, 0x4d, 0xbd,
+        0x57, 0x5d, 0x78, 0xba, 0x87, 0xd4, 0xa4, 0xbc,
+        0x47, 0x46, 0x66, 0x6e, 0x7a, 0x4e, 0x6e, 0xe1,
+        0x57, 0x2b, 0xbf, 0xfc, 0x2e, 0xba, 0x30, 0x8a,
+        0x2d, 0x82, 0x5c, 0xb7, 0xb4, 0x1f, 0xde, 0x3a,
+        0x95, 0xd5, 0xcf, 0xf0, 0xdf, 0xa2, 0xd0, 0xfd,
+        0xd6, 0x36, 0xb3, 0x2a, 0xea, 0x8b, 0x4a, 0x3c,
+        0x53, 0x27, 0x42, 0xd3, 0x30, 0xbd, 0x1b, 0x90
+    };
+
+    p = params3;
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_PASSWORD,
+                                             (char *)"1234567890", 10);
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SALT,
+                                             (char *)"saltsalt", 8);
+    *p++ = OSSL_PARAM_construct_int(OSSL_KDF_PARAM_THREADS, &threads);
+    *p++ = OSSL_PARAM_construct_int(OSSL_KDF_PARAM_ARGON2_LANES, &threads);
+    *p++ = OSSL_PARAM_construct_int(OSSL_KDF_PARAM_ARGON2_MEMCOST, &m_cost);
+    *p = OSSL_PARAM_construct_end();
+
+    ret =
+        TEST_ptr(kctx = get_kdfbyname(SN_argon2d))
+        && TEST_true(EVP_KDF_set_ctx_params(kctx, params3))
+        && TEST_int_eq(EVP_KDF_derive(kctx, out3, sizeof(out3)), 1)
+        && TEST_mem_eq(out3, sizeof(out3), expected_res3, sizeof(expected_res3));
+
+    EVP_KDF_free_ctx(kctx);
+
+    if (ret == 0)
+        return 0;
+
+    /* Test m_cost >= 8 lanes constraint. */
+
+    m_cost = 10;
+    threads = 2;
+    p = params4;
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_PASSWORD,
+                                             (char *)"1234567890", 10);
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SALT,
+                                             (char *)"saltsalt", 8);
+    *p++ = OSSL_PARAM_construct_int(OSSL_KDF_PARAM_THREADS, &threads);
+    *p++ = OSSL_PARAM_construct_int(OSSL_KDF_PARAM_ARGON2_LANES, &threads);
+    *p++ = OSSL_PARAM_construct_int(OSSL_KDF_PARAM_ARGON2_MEMCOST, &m_cost);
+    *p = OSSL_PARAM_construct_end();
```

Comment:
Its a bit messy lumping all 3 tests in here..
It may be cleaner to do them as seperate tests.

You can also use ADD_ALL_TESTS and data drive it..

Reply:
Split it into multiple functions -- test cases.